### PR TITLE
Apply: Migration from CSA to SSA; set operation and merge

### LINKED ISSUE #14520: client-go applier fails to upgrade resources
The addon upgrade test has been failing since we migrated to the client-go-based applier in #14362: https://testgrid.k8s.io/kops-misc#kops-aws-addon-resource-tracking

This needs to be looked at or we need to prepare for flipping back to kubectl for another release.

/cc @justinsb

### PR DESCRIPTION
**Pull request by justinsb**
If we don't, the (kops,Update) managed blocks are treated as different
from the (kops,Apply) block.  This causes conflicts when the two
blocks specific incompatible values.

We also need to merge the field managers to merge fields that (now) have the same key.

Issue #14520 (I believe it is the fix!)

### CODE REVIEW

File: pkg/applylib/applyset/applyset.go [hunk 1]
```diff
@@ -191,6 +191,11 @@ func createManagedFieldPatch(currentObject *unstructured.Unstructured) ([]byte,
 		fixedManagedField := managedField.DeepCopy()
 		if managedField.Manager == "kubectl-edit" || managedField.Manager == "kubectl-client-side-apply" {
 			fixedManagedField.Manager = "kops"
+			fixedManagedField.Operation = "Apply"
+		}
+		// In case we have a kops & Update manager
+		if fixedManagedField.Manager == "kops" && fixedManagedField.Operation == "Update" {
```
[the remark this answers is not in the document]
Comment (justinsb) [reply pkg/applylib/applyset/applyset.go]: I added some tests and it looks like the server doesn't combine them for us.  So I added code to merge the field managers.  It does require a dependency on apiserver, but this is the only place it's exposed (currently) and it's supposed to be a relatively stable library.

**Comment by olemarkus** [comment]
From what I can tell, this looks good. But there is a lot of code being added without unit tests. Can we add some before this merges?

**Comment by justinsb** [comment]
I was actually keeping the test in a separate PR in case we wanted to cherry-pick: https://github.com/kubernetes/kops/pull/14623 . Some problems though:

* Maybe we don't care about cherry-picking, the tests aren't that big.  In that case I can just merge those tests in to this PR.
* They aren't unit tests, because they need a real k8s apiserver.  kind suffices, but e.g. [mockkubeapiserver](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern) likely does not.  There's also envtest from controller-runtime / kubebuilder.  The test is skipped unless an env var is present (that works better than build tags with e.g. gopls), so it doesn't add new dependencies to a normal test run, just it's a new dependency to run the tests.

What I could do is add a prow job or github action job, that would run with a kind cluster?

**Comment by olemarkus** [comment]
I am thinking that a lot of the code doesn't interact with the server so it would be possible to create unit tests for them. But I may be missing something here.

**Comment by justinsb** [comment]
Ah - excellent point @olemarkus .  I can certainly create a unit test for the Migrator (UnstructuredClient is really just a refactor, and arguably we could just use the version from controller-runtime, but I think that pulls in some extra dependencies).

The values will be a little "magic" for the unit tests - being derived from "real kube-apiservers", but I think that's fine for a unit test.  Good idea!

**Comment by justinsb** [comment]
Added some unit tests - thanks for the suggestion @olemarkus !

**Review by olemarkus — APPROVED** [review]
This is great. Thanks!

**Pull request merged**